pkg/diff/diff.go

```diff
@@ -283,6 +285,29 @@ func removeWebhookMutation(predictedLive, live *unstructured.Unstructured, gvkPa
 	return &unstructured.Unstructured{Object: pl}, nil
 }
 
+// appendKeyFields appends the key fields like "name" to the set elements which have some entries keyed by those fields.
+// This is needed to merge properly, see https://github.com/argoproj/argo-cd/issues/20792.
+func appendKeyFields(set *fieldpath.Set) {
+	keyFieldPaths := []fieldpath.Path{}
+	set.Iterate(func(path fieldpath.Path) {
+		for i := 0; i < len(path); i++ {
+			if path[i].Key != nil {
+				for _, keyField := range *path[i].Key {
+					pathPartCopy := make([]fieldpath.PathElement, i+1)
+					copy(pathPartCopy, path[:i+1])
+					newPath := append(pathPartCopy, fieldpath.PathElement{FieldName: &keyField.Name})
```

Comment:
You are assuming that the field `name` is always the key in lists which can be true or false depending on the resource. We can't assume schema structure from this library and must rely on the provided `GvkParser`. The problem is likely cause by the fact that the `GvkParser` isn't populated with the necessary default values and list keys for the specific schema. It is strange, however that the error is being reported from a `Deployment` resource as this should contain the necessary schema info IRRC. 
Please investigate more and let me know your findings.

Reply:
@andrii-korotkov-verkada I understand. My point is, how you determine what is the key that should be considered when merging 2 hypothetical lists?

Reply:
All of the keys should be considered as far as I understand. Let me take a further look to see if we can derive from schema.

Reply:
I also don't know if key fields are supposed to be returned by the diff, e.g. since they aren't necessarily removed. It's just we use those fields in a special way to do a merge, which has a different requirements that diff computation can have.

Reply:
For example, we don't want key fields in added items, since we remove them in this logic. So not sure default behavior of the libs should be changed and it's on us to adjust, which I'm trying to do with this PR.